### PR TITLE
feat(opentelemetry): add opentelemetry variables

### PR DESCRIPTION
**Pull request by lework**
### Description

Span information recorded in the access log.

Fixes #  https://github.com/apache/apisix/issues/8765

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [x] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

**Comment by lework** [comment]
@starsz  Please take a look when you have free time, thanks!

**Comment by membphis** [comment]
missing test cases @lework

File: docs/en/latest/plugins/opentelemetry.md [hunk 1]
```diff
@@ -89,6 +89,32 @@ plugin_attr:
...
+- `opentelemetry_span_id` -  Span Id of the current span
+
+How to use variables? you have to add it to your configuration file (`conf/config.yaml`):
+
+```yaml title="./conf/config.yaml"
+    nginx_config:
+      http_server_configuration_snippet: |
+        set $opentelemetry_context_traceparent ""
+        set $opentelemetry_trace_id ""
+        set $opentelemetry_span_id ""
```
Comment (starsz) [inline docs/en/latest/plugins/opentelemetry.md:107]: I think there is another way, that if we enable `OpenTelemetry` we can add `opentelemetry_span_id` automatically in `ngx_tpl.lua`.

File: docs/en/latest/plugins/opentelemetry.md [hunk 2]
```diff
@@ -89,6 +89,32 @@ plugin_attr:
...
+- `opentelemetry_context_traceparent` -  [W3C trace context](https://www.w3.org/TR/trace-context/#trace-context-http-headers-format), e.g.: `00-0af7651916cd43dd8448eb211c80319c-b9c7c989f97918e1-01`
+- `opentelemetry_trace_id` - Trace Id of the current span
+- `opentelemetry_span_id` -  Span Id of the current span
+
+How to use variables? you have to add it to your configuration file (`conf/config.yaml`):
+
+```yaml title="./conf/config.yaml"
+    nginx_config:
+      http_server_configuration_snippet: |
+        set $opentelemetry_context_traceparent ""
```
Comment (starsz) [inline docs/en/latest/plugins/opentelemetry.md:105]: Hi, I think there is another way that we can add those lines in `cli/ngx_tpl.lua` automatically if we enabled the "OpenTelemetry" plugins.

Else APISIX will panic the request is we forget to set those line but enable the `set_ngx_var` options in the plugin.

Comment (lework) [reply docs/en/latest/plugins/opentelemetry.md]: Good idea, I added some code

**Comment by lework** [comment]
@starsz That ci error above looks like an error reported by another plugin, what should I do?

**Comment by lework** [comment]
@membphis  Looking for help on what I should do next?

**Review by starsz — DISMISSED** [review]

**Comment by moonming** [comment]
@monkeyDluffy6017 please review this PR

**Comment by lework** [comment]
@monkeyDluffy6017 please review this PR

**Comment by moonming** [comment]
@monkeyDluffy6017 any update？

**Comment by monkeyDluffy6017** [comment]
Nice work!

File: apisix/plugins/opentelemetry.lua [hunk 3]
```diff
@@ -112,6 +113,11 @@ local attr_schema = {
             },
             default = {},
         },
+        set_ngx_var = {
+          type = "boolean",
+          description = "set nginx variables",
+          default = false,
+        },
```
Comment (leslie-tsang) [inline apisix/plugins/opentelemetry.lua:120]:
```suggestion
        set_ngx_var = {
            type = "boolean",
            description = "set nginx variables",
            default = false,
        },
```

File: apisix/plugins/opentelemetry.lua [hunk 4]
```diff
@@ -332,6 +338,17 @@ function _M.rewrite(conf, api_ctx)
...
     })
+
+    if plugin_info.set_ngx_var then
+      local span_context = ctx:span():context()
+      ngx_var.opentelemetry_context_traceparent = string_format("00-%s-%s-%02x",
+                                                                 span_context.trace_id,
+                                                                 span_context.span_id,
+                                                                 span_context.trace_flags)
+      ngx_var.opentelemetry_trace_id = span_context.trace_id
+      ngx_var.opentelemetry_span_id = span_context.span_id
```
Comment (leslie-tsang) [inline apisix/plugins/opentelemetry.lua:349]: Would be better to use `ctx.var` instead ?

**Review by leslie-tsang — APPROVED** [review]
The rest LGTM.

**Comment by moonming** [comment]
@lework, thanks for your contribution 👍
@indrekj Welcome to continue migrating to Apache APISIX

**Pull request merged**